### ISSUE TITLE
Global setting "Ignore unverified SSL certificates" without effect

### ISSUE THREAD

**Issue by johnjohndoe** [jenkinsci/stashnotifier-plugin#179]
I noticed that the setting `Ignore unverified SSL certificates` in Jenkins' system configuration is ignored.

# Environment
- Jenkins 1.651.3
- Stash Notifier 1.12
- Self-hosted Stash instance (version unclear)
- `Ignore unverified SSL certificates` enabled in system settings
- `Ignore unverified SSL certificates` disabled in job settings

# Failure

The console log shows the following output for the post-build task:

``` java
Notifying Stash at "https://stash.exmaple.com"
Caught exception while notifying Stash with id 5e150ebf30e3c6fed1d90aff80b21c2ee9bc03fa
javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1949)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:302)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:914)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
	at org.apache.http.impl.conn.BasicHttpClientConnectionManager.connect(BasicHttpClientConnectionManager.java:338)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at org.jenkinsci.plugins.stashNotifier.StashNotifier.notifyStash(StashNotifier.java:726)
	at org.jenkinsci.plugins.stashNotifier.StashNotifier.processJenkinsEvent(StashNotifier.java:305)
	at org.jenkinsci.plugins.stashNotifier.StashNotifier.perform(StashNotifier.java:258)
	at org.jenkinsci.plugins.stashNotifier.StashNotifier.perform(StashNotifier.java:218)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:387)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
	at sun.security.validator.Validator.validate(Validator.java:260)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1491)
	... 33 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:146)
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:131)
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
	... 39 more
)
Email was triggered for: Failure - Any
Sending email for trigger: Failure - Any
An attempt to send an e-mail to empty list of recipients, ignored.
Finished: FAILURE
```

# Workaround
- Enabling the **same** setting **in a specific job** let's the request finish successfully.

```
Notifying Stash at "https://stash.example.com"
Notified Stash for commit with id 1c03cb834ef1b3bc473b3d2f5981f8b66d0f78f3
```

**Comment by scaytrase** [comment]
Yeah, looks like a weird bug. Both flags are checked but only project level flag is passed to ssl connection factory. 

https://github.com/jenkinsci/stashnotifier-plugin/blob/e13a45309a0b70406c3556725ace743fdc91c77b/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L401-L422

**Comment by johnjohndoe** [comment]
@scaytrase I changed the error message in the issue above. I pasted the wrong one last week.

---

I started an attempt to solve this issue but _findbugs_ tells me that the condition is useless:

> [INFO] Useless condition: it's known that ignoreUnverifiedSSL == true at this point [org.jenkinsci.plugins.stashNotifier.StashNotifier, org.jenkinsci.plugins.stashNotifier.StashNotifier] Unreachable code at StashNotifier.java:[line 416]At StashNotifier.java:[line 416] UC_USELESS_CONDITION

**Comment by scaytrase** [comment]
Sure, *findbugs* is correct as always. On [line 416](https://github.com/johnjohndoe/stashnotifier-plugin/blob/872fd488c7588026498299b0f908c12b6ad8d610/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L416-L421) you have the start of the method call, one of parameters is the expression

```java
ignoreUnverifiedSSL ? SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER : null
```

But you reach this line only if 

```java
url.getProtocol().equals("https") && ignoreUnverifiedSSL
```

evalates to `true`, which means that right side is also have to be `true`.

Which means ternary operator becomes useless and could be replaced with it's true branch value.

https://github.com/johnjohndoe/stashnotifier-plugin/blob/872fd488c7588026498299b0f908c12b6ad8d610/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L416-L421

**Comment by johnjohndoe** [comment]
Yes, I saw this from the _findbugs_ output. But still you want to check **both** the global and the job specific setting, right? Which is the jobs specific setting - the one derived from `DescriptorImpl`?

**Comment by scaytrase** [comment]
`DescriptorImpl` is the global config AFAIK. Checking both ( `any of`) is used with `||` expression, so any of checked options will allow to skip TLS verification

**Comment by johnjohndoe** [comment]
Agreed. I am checking the `||` expression [here](https://github.com/johnjohndoe/stashnotifier-plugin/commit/872fd488c7588026498299b0f908c12b6ad8d610#diff-9905a5d95ec4490fc08dccacee897bc2R410). But how can I make _findbugs_ happy?

**Comment by scaytrase** [comment]
> Which means ternary operator becomes useless and could be replaced with it's true branch value.

This should help

**Comment by johnjohndoe** [comment]
Sorry, but I really don't get it. We want to evaluate both the global and the job specific setting. Why should I remove the check?

**Comment by scaytrase** [comment]
I'm speaking about this one
```java
 ignoreUnverifiedSSL ? SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER : null 
```

`ignoreUnverifiedSSL` is always true inside the `if`

**Comment by johnjohndoe** [comment]
But that is incorrect - isn't it?

- `ignoreUnverifiedSSLPeer` is set via the [contructor](https://github.com/johnjohndoe/stashnotifier-plugin/blob/872fd488c7588026498299b0f908c12b6ad8d610/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L149)
- `descriptor.isIgnoreUnverifiedSsl()` is defined by the `DescriptorImpl`, which is the global setting according to what @scaytrase wrote before.

So both settings can be either `true` or `false` during runtime. Or am I missing something?

**Comment by scaytrase** [comment]
I think you mixing local var `ignoreUnverifiedSSL` (set as `||` result) and property `ignoreUnverifiedSSLPeer`

**Comment by johnjohndoe** [comment]
Pleae correct me if I am wrong:

- `ignoreUnverifiedSSLPeer` is a property of `StashNotifier` assignable via the `@DataBoundConstructor` which represents the job specific setting.
- `ignoreUnverifiedSSL` is the boolean result of `ignoreUnverifiedSSLPeer || descriptor.isIgnoreUnverifiedSsl();`
- `descriptor.isIgnoreUnverifiedSsl()` is defined by the `DescriptorImpl`, which is the global setting according to what @scaytrase wrote before.

**Comment by scaytrase** [comment]
Looks weird but correct. Bad namings are so bad.

**Comment by johnjohndoe** [comment]
It would also add to the readability if `DescriptorImpl` is moved into it's own file.

How do we continue on this issue?

**Comment by scaytrase** [comment]
This class requires heavy refactor in many terms. I think ATM we need minimal patch to fix issue. Refactoring patches are also welcome as separare PR, it would be easier to review and test

**Comment by johnjohndoe** [comment]
I agree. The refactoring should be on a separate PR.

For this issue I cannot figure out the fix it seems.

**Comment by johnjohndoe** [comment]
@scaytrase I rebased my branch and force pushed it up.

I tested my changes on my local Jenkins instance but **still global settings are ignored**.
Can you please help me to resolve the issue - I think I am missing something here. I can create a pull request if you prefer to comment on that - please let me know.

**Comment by scaytrase** [comment]
@johnjohndoe please  feel free to open a PR, this will make code discussion a bit easier

**Comment by johnjohndoe** [comment]
@scaytrase When will this be released?

**Comment by scaytrase** [comment]
wow, sorry, I've thought that I release all pending patches last time. I'll release latest version on this weekend then.

Thank you for pinging me for that

**Comment by scaytrase** [comment]
@johnjohndoe I've published 1.14 just now
